src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/CosmosFhirDataStore.cs

```diff
@@ -108,7 +108,44 @@ public sealed class CosmosFhirDataStore : IFhirDataStore, IProvideCapability
             _modelInfoProvider = modelInfoProvider;
         }
 
-        public async Task<UpsertOutcome> UpsertAsync(
+        public async Task<IDictionary<ResourceKey, UpsertOutcome>> MergeAsync(IReadOnlyList<ResourceWrapperOperation> resources, CancellationToken cancellationToken)
+        {
+            var results = new Dictionary<ResourceKey, UpsertOutcome>();
+            if (resources == null || resources.Count == 0)
+            {
+                return results;
+            }
+
+            foreach (var resource in resources)
+            {
+                results.Add(resource.Wrapper.ToResourceKey(), await UpsertAsync(resource, cancellationToken));
+            }
+
+            return results;
+        }
+
+        public async Task<IReadOnlyList<ResourceWrapper>> GetAsync(IReadOnlyList<ResourceKey> keys, CancellationToken cancellationToken)
+        {
+            var results = new List<ResourceWrapper>();
+            if (keys == null || keys.Count == 0)
+            {
+                return results;
+            }
+
+            foreach (var key in keys)
+            {
+                results.Add(await GetAsync(key, cancellationToken));
+            }
```

Comment:
Guessing the syntax here, but why not use a parallel select?

```suggestion
            results.AddRange(await Task.WhenAll(keys.Select(async key => await GetAsync(key, cancellationToken))));
```

Reply:
changed